### PR TITLE
ssh/tailssh: monitor IPNBus for recorder host updates

### PR DESCRIPTION
**Pull request by patrickod**
Monitor the IPNBus for netmap updates that indicate that the previously selected SSH recorder node has become inaccessible and proactively terminate the SSH session if this happens.

Where possible set tighter KeepAlive parameters for the underlying connection to the recorder host.

Updates tailscale/corp#24023

### CODE REVIEW

File: sessionrecording/connect.go [hunk 1]
```diff
@@ -130,8 +130,23 @@ func SessionRecordingClientForDialer(dialCtx context.Context, dial func(context.
 				cancel()
 			}
 		}()
-		return dial(perAttemptCtx, network, addr)
+
+		conn, err := dial(perAttemptCtx, network, addr)
```
Comment (awly) [inline sessionrecording/connect.go:134]: let's make this a separate PR, once we prove that these keepalives actually work

File: ssh/tailssh/tailssh.go [hunk 2]
```diff
@@ -1517,8 +1522,58 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
...
+
+			var mask ipn.NotifyWatchOpt
+			// identify the recorder node key by its tailscale IP
+			status := ss.conn.srv.lb.Status()
+			var recorderNodeID key.NodePublic
+			if status != nil && len(status.Peers()) > 0 {
+			outer:
+				for nodeID, peer := range status.Peer {
+					for _, ip := range peer.TailscaleIPs {
+						if ip == last.Recorder.Addr() {
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1538]: nit: this can be `slices.Contains(peer.TailscaleIPs, last.Recorder.Addr()`

File: ssh/tailssh/tailssh.go [hunk 3]
```diff
@@ -1517,8 +1522,58 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
 			ss.logf("recording: error starting recording (failing open): %v", err)
 			return nil, nil
 		}
+
+		recorderDisappear := make(chan struct{})
+		go func() {
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:1527]: can you promote this to its own named method? Both to reduce the size/indentation of the outer method here, and because we try to avoid unnamed long-lived goroutines, as that slows debugging reading stack traces from goroutine dumps.

File: ssh/tailssh/tailssh.go [hunk 4]
```diff
@@ -1517,8 +1522,58 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
...
+				for _, peer := range roNotify.NetMap.Peers {
+					if peer.Key() == recorderNodeID {
+						recorderPeer = peer
+						break
+					}
+				}
+
+				if !recorderPeer.Valid() || !*recorderPeer.Online() {
+					ss.logf("previously selected recorder node is now offline or invalid; terminating session")
+					recorderDisappear <- struct{}{}
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1563]: to avoid a goroutine leak if `errChan` fires first:
```suggestion
					close(recorderDisappear)
```

File: ssh/tailssh/tailssh.go [hunk 5]
```diff
@@ -1517,8 +1522,58 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
...
+				}
+			}
+			// if we can't find the recorder error out.
+			if recorderNodeID.IsZero() {
+				ss.logf("unable to find recorder node in status peer list")
+				return
+			}
+
+			// watch for IPN updates to observe if the recorder node ever goes offline.
+			ss.conn.srv.lb.WatchNotificationsAs(context.Background(), nil, mask, func() {}, func(roNotify *ipn.Notify) bool {
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1552]: we should terminate this when the session ends. you can use `ss.ctx` instead of `context.Background()` to catch that.

File: ssh/tailssh/tailssh.go [hunk 6]
```diff
@@ -1517,8 +1522,58 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
...
 		}
+
+		recorderDisappear := make(chan struct{})
+		go func() {
+			last := attempts[len(attempts)-1]
+
+			var mask ipn.NotifyWatchOpt
+			// identify the recorder node key by its tailscale IP
+			status := ss.conn.srv.lb.Status()
+			var recorderNodeID key.NodePublic
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1533]: nit:
```suggestion
			var recorderNodeKey key.NodePublic
```

File: ssh/tailssh/tailssh.go [hunk 7]
```diff
@@ -1517,8 +1522,58 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
...
+				var recorderPeer tailcfg.NodeView
+				for _, peer := range roNotify.NetMap.Peers {
+					if peer.Key() == recorderNodeID {
+						recorderPeer = peer
+						break
+					}
+				}
+
+				if !recorderPeer.Valid() || !*recorderPeer.Online() {
+					ss.logf("previously selected recorder node is now offline or invalid; terminating session")
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1562]: nit: include the recorder node ID and IP in the error message

File: ssh/tailssh/tailssh.go (hunk 3 above)
Comment (patrickod) [reply ssh/tailssh/tailssh.go]: fair shout. done

**Review by bradfitz — COMMENTED** [review]
Thinking about this a bit more, I don't think IPNBus is the route we want to take to monitor peer reachability.

In addition to depending on control to send the messages (which are often delayed/batched/filtered/lost), we should actively probe the recorder. We already have a connection open to it, so we should have an every-N-second heartbeat going to it (e.g. use http2.Transport's built-in-one).

We don't do that in the general case for the happy green dot because we don't want to actively probe, but in this case we're already in communication with the recorder (and we're a server thats somebody's SSHing into, so not on battery probably) so a few more bytes of actively probing the recorder server seems more reliable and lower latency.

File: ssh/tailssh/tailssh.go [hunk 8]
```diff
@@ -1448,6 +1454,53 @@ func (ss *sshSession) openFileForRecording(now time.Time) (_ io.WriteCloser, err
...
 }
 
+// monitorRecorderStatus watches the IPNBus for any netmap updates that indicate
+// that the recorder node is now inaccessible and signals via the returned
+// channel.
+func (ss *sshSession) monitorRecorderStatus(last *tailcfg.SSHRecordingAttempt) (<-chan struct{}, error) {
+	// identify the recorder node key by its tailscale IP
+	status := ss.conn.srv.lb.Status()
+	var recorderNodeKey key.NodePublic
+	if status != nil && len(status.Peers()) > 0 {
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1464]:
```suggestion
	if status != nil && len(status.Peer) > 0 {
```

File: ssh/tailssh/tailssh.go [hunk 9]
```diff
@@ -1448,6 +1454,53 @@ func (ss *sshSession) openFileForRecording(now time.Time) (_ io.WriteCloser, err
...
+	if recorderNodeKey.IsZero() {
+		ss.logf("unable to find recorder node in status peer list")
+		return nil, errors.New("recorder node not found")
+	}
+
+	// watch indefinitely for IPN updates to observe if the recorder node ever goes offline.
+	recorderOffline := make(chan struct{})
+	go func() {
+		var mask ipn.NotifyWatchOpt
+		ss.conn.srv.lb.WatchNotificationsAs(ss.ctx, nil, mask, func() {}, func(roNotify *ipn.Notify) bool {
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1482]: optional, these are equivalent
```suggestion
		ss.conn.srv.lb.WatchNotifications(ss.ctx, mask, func() {}, func(roNotify *ipn.Notify) bool {
```

File: ssh/tailssh/tailssh.go [hunk 10]
```diff
@@ -1448,6 +1454,53 @@ func (ss *sshSession) openFileForRecording(now time.Time) (_ io.WriteCloser, err
...
+			var recorderPeer tailcfg.NodeView
+			for _, peer := range roNotify.NetMap.Peers {
+				if peer.Key() == recorderNodeKey {
+					recorderPeer = peer
+					break
+				}
+			}
+
+			if !recorderPeer.Valid() || !*recorderPeer.Online() {
+				ss.logf("previously selected recorder node %q at %d is now offline or invalid; terminating session", recorderNodeKey, last.Recorder.Addr())
```
Comment (awly) [inline ssh/tailssh/tailssh.go:1492]:
```suggestion
				ss.logf("previously selected recorder node %q at %q is now offline or invalid; terminating session", recorderNodeKey, last.Recorder.Addr())
```

File: ssh/tailssh/tailssh.go (hunk 9 above)
Comment (awly) [inline ssh/tailssh/tailssh.go:1482]:
```suggestion
		ss.conn.srv.lb.WatchNotificationsAs(ss.ctx, nil, mask, func() {}, func(roNotify *ipn.Notify) bool {
			if roNotify.NetMap == nil {
				return true
			}
```

**Review by awly — COMMENTED** [review]
Tested and it works with the tweaks in the review.
Could you also take a stab at writing a unit test? There's some plumbing in https://github.com/tailscale/tailscale/blob/78103f8e99bf46cb053ea74d8cf0e04702fecaed/ssh/tailssh/tailssh_test.go#L489, you just have to inject some notifications via `localState` in there.

**Review by bradfitz — CHANGES_REQUESTED** [review]
Let's finish the design discussion (currently on Slack) before merging this. I really don't think this is where we want to go.
